controllers/machinepool_controller_phases.go

```diff
@@ -0,0 +1,260 @@
+	patchHelper, err := patch.NewHelper(obj, r.Client)
+	if err != nil {
+		return external.ReconcileOutput{}, err
+	}
+
+	// Set external object OwnerReference to the MachinePool.
+	ownerRef := metav1.OwnerReference{
+		APIVersion: clusterv1.GroupVersion.String(),
+		Kind:       "MachinePool",
+		Name:       m.Name,
+		UID:        m.UID,
+	}
+
+	// Add ownerRef to object.
+	obj.SetOwnerReferences(util.EnsureOwnerRef(obj.GetOwnerReferences(), ownerRef))
+
+	// Set the Cluster label.
+	labels := obj.GetLabels()
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+	labels[clusterv1.ClusterLabelName] = m.Spec.ClusterName
+	obj.SetLabels(labels)
+
+	// Always attempt to Patch the external object.
+	if err := patchHelper.Patch(ctx, obj); err != nil {
+		return external.ReconcileOutput{}, err
+	}
+
+	// Add watcher for external object, if there isn't one already.
+	_, loaded := r.externalWatchers.LoadOrStore(obj.GroupVersionKind().String(), struct{}{})
+	if !loaded && r.controller != nil {
+		logger.Info("Adding watcher on external object", "gvk", obj.GroupVersionKind())
+		err := r.controller.Watch(
+			&source.Kind{Type: obj},
+			&handler.EnqueueRequestForOwner{OwnerType: &clusterv1.MachinePool{}},
+		)
+		if err != nil {
+			r.externalWatchers.Delete(obj.GroupVersionKind().String())
+			return external.ReconcileOutput{}, errors.Wrapf(err, "failed to add watcher on external object %q", obj.GroupVersionKind())
+		}
+	}
+
+	// Set failure reason and message, if any.
+	failureReason, failureMessage, err := external.FailuresFrom(obj)
+	if err != nil {
+		return external.ReconcileOutput{}, err
+	}
+	if failureReason != "" {
+		machineStatusFailure := capierrors.MachinePoolStatusFailure(failureReason)
+		m.Status.FailureReason = &machineStatusFailure
+	}
+	if failureMessage != "" {
+		m.Status.FailureMessage = pointer.StringPtr(
+			fmt.Sprintf("Failure detected from referenced resource %v with name %q: %s",
+				obj.GroupVersionKind(), obj.GetName(), failureMessage),
+		)
+	}
+
+	return external.ReconcileOutput{Result: obj}, nil
+}
+
+// reconcileBootstrap reconciles the Spec.Bootstrap.ConfigRef object on a MachinePool.
+func (r *MachinePoolReconciler) reconcileBootstrap(ctx context.Context, cluster *clusterv1.Cluster, m *clusterv1.MachinePool) error {
+	// Call generic external reconciler if we have an external reference.
+	var bootstrapConfig *unstructured.Unstructured
+	if m.Spec.Template.Spec.Bootstrap.ConfigRef != nil {
+		bootstrapReconcileResult, err := r.reconcileExternal(ctx, cluster, m, m.Spec.Template.Spec.Bootstrap.ConfigRef)
+		if err != nil {
+			return err
+		}
+		// if the external object is paused, return without any further processing
+		if bootstrapReconcileResult.Paused {
+			return nil
+		}
+		bootstrapConfig = bootstrapReconcileResult.Result
+	}
+
+	// If the bootstrap data is populated, set ready and return.
+	if m.Spec.Template.Spec.Bootstrap.Data != nil || m.Spec.Template.Spec.Bootstrap.DataSecretName != nil {
```

Comment:
@vincepri do we want to be looking at .data?

Reply:
We should only look at the secret